### ISSUE TITLE
aws s3 sync doesn't guess MIME type for .woff file

### ISSUE THREAD

**Issue by Janpot** [aws/aws-cli#1249]
I'm syncing an assets folder to S3 to be served over CDN. I noticed that .woff files get `Content-Type: binary/octet-stream` instead of `application/font-woff`.

**Comment by varmint09** [comment]
+1 for this.

**Comment by azoff** [comment]
+1

**Comment by j** [comment]
+1 (and for other font types)

**Comment by azoff** [comment]
For now I'm using this: https://github.com/dailymuse/s3-content-type-fixer

**Comment by wilsonwc** [comment]
This looks like it's actually a limitation of the Python mimetype library, maybe the code needs to be augmented to support mime types that aren't defined in that library?

**Comment by jamesls** [comment]
Yes, and internally I believe it will look at `/etc/mime.types` and a few other locations to get these mappings, so a temporary workaround would be for people to update that file if possible.

Otherwise I think there's two options here:

1. We can either start an additional list within the AWS CLI code based that has additional extensions->mimetype mappings.
2. We can look for and load an optional [mime.types](https://wiki.debian.org/MIME/etc/mime.types) file.  Say if you were to create an `~/.aws/cli/mime.types` file, we would load that file and use that in addition to `/etc/mime.types`.

Any preferences for either approach?

**Comment by wilsonwc** [comment]
Thanks for the reply.  It looks like Apache maintains an updated list that would be the same format as the /etc/mime.types. This might be appropriate to link to or include to satisfy the first option, and then still give users the ability to add/override types in a ~/.aws/cli/mime.types file as a new feature.

http://svn.apache.org/viewvc/httpd/httpd/trunk/docs/conf/mime.types?view=markup

**Comment by wilsonwc** [comment]
I can also confirm that replacing my /etc/mime.types with the file linked above fixed the issue for me as a temporary workaround.

**Comment by dbinetti** [comment]
@jamesls @wilsonwc Are there any plans to to add support for the `~/.aws/cli/mime.types` approach?  This would be most welcome.

**Comment by ASayre** [comment]
Good Morning!

We're closing this issue here on GitHub, as part of our migration to [UserVoice](https://aws.uservoice.com/forums/598381-aws-command-line-interface) for feature requests involving the AWS CLI.

This will let us get the most important features to you, by making it easier to search for and show support for the features you care the most about, without diluting the conversation with bug reports.

As a quick UserVoice primer (if not already familiar): after an idea is posted, people can vote on the ideas, and the product team will be responding directly to the most popular suggestions.

We’ve imported existing feature requests from GitHub - Search for this issue there!

And don't worry, this issue will still exist on GitHub for posterity's sake.  As it’s a text-only import of the original post into UserVoice, we’ll still be keeping in mind the comments and discussion that already exist here on the GitHub issue.

GitHub will remain the channel for reporting bugs.  

Once again, this issue can now be found by searching for the title on: https://aws.uservoice.com/forums/598381-aws-command-line-interface 

-The AWS SDKs & Tools Team

**Comment by jamesls** [comment]
Based on community feedback, we have decided to return feature requests to GitHub issues.

**Comment by veloper** [comment]
For those using Ubuntu (at least 14.04) it looks like you can add a personalized list of mime types to your home directory that will override the system's `/etc/mime.types` file.

Per the docs in the `/etc/mime.types` file:

    #  Users can add their own types if they wish by creating a ".mime.types"
    #  file in their home directory.  Definitions included there will take
    #  precedence over those listed here.

So, you can script some individual replacements like...

    echo -e "font/woff woff" >> ~/.mime.types
    echo -e "font/woff2 woff2" >> ~/.mime.types

Or, just do the whole _kit-and-caboodle_ wrapping @wilsonwc 's solution...

    curl -s "http://svn.apache.org/viewvc/httpd/httpd/trunk/docs/conf/mime.types?revision=1810121&view=co" > ~/.mime.types

Happy coding & orchestrating!

**Comment by 75lb** [comment]
On my macbook, the `aws s3 sync` command incorrectly guesses that an `index.mjs` file should have the MIME type `binary/octet-stream`. This breaks my web app - the correct MIME type should be `application/javascript`. 

What is the currently recommended way to override the MIME type guessed by `aws s3 sync` on a mac?

**Comment by lucashfreitas** [comment]
I am using the latest version of aws cli in Macbook High Sierra and having same issue!Currently S3 is classifying .woff and .woff2 files with html content type. Anyone find any workaround for that in Macbook?

**Comment by JJC1138** [comment]
For me on Ubuntu 18.04.1 LTS `aws s3 sync` didn't make use overrides that I put in `~/.mime.types`, but did make use of ones that I put in the system-wide `/etc/mime.types` using e.g:

``` echo "video/mp4 m4v" | sudo tee -a /etc/mime.types```

**Comment by odyth** [comment]
For those on windows aws cli pulls the content type from HKEY_LOCAL_MACHINE\Software\Classes\.js and for me on windows server 2016 and windows 10 this was defaulted to text/plain. Changing HKEY_LOCAL_MACHINE\Software\Classes\.js to have a content type of application/javascript resolved the issue for me.

**Comment by clearwaterstream** [comment]
> 
> 
> For those on windows aws cli pulls the content type from HKEY_LOCAL_MACHINE\Software\Classes.js and for me on windows server 2016 and windows 10 this was defaulted to text/plain. Changing HKEY_LOCAL_MACHINE\Software\Classes.js to have a content type of application/javascript resolved the issue for me.

Thank you. I had to do it for javascript and the woff files. I am not sure if AWS sync tool should be modifying those registry keys on Windows. Perhaps and it could, like on Linux system, pull the info from mime.types file. Since that file on Windows does not exist, upon first run the tool can create it, taking cues from the registry ...

**Comment by caioproiete** [comment]
I'm seeing this same issue using `aws s3 sync` on Windows, where Javascript files are not being set with the correct mime type which makes scripts to be blocked by browsers.

I double-checked the Registry:
- `HKEY_CLASSES_ROOT\.js`
- `HKEY_LOCAL_MACHINE\Software\Classes\.js`

And in both cases the `Content Type` value is set to `application/javascript`, yet this mime type is not set in the S3 bucket.

**Comment by liesislukas** [comment]
/etc/mime.types has font/woff mime type:

![image](https://user-images.githubusercontent.com/2733862/112991103-71b4be00-916f-11eb-8571-4b4fd85cbed9.png)

using code build to build the app artifact is uploaded to s3 by aws itself

![image](https://user-images.githubusercontent.com/2733862/112991206-8b560580-916f-11eb-8036-955749167f07.png)

yet inside s3 mime type is broken. file `fa-solid-900.woff`

![image](https://user-images.githubusercontent.com/2733862/112991351-b17ba580-916f-11eb-9cc5-9a30de8ab4b0.png)

**Comment by eboboshka** [comment]
It's very strange that the error has remained relevant since 2015! Used AWS S3 CLI to just copy static landing pages into S3 and kept getting this weird problem with webp & ttf files (binary/octet-stream)

This adds even more questions that the AWS client didn't add normal MIME handling, since AWS CloudFront is very demanding on Content-Type for compression to work properly - https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/ServingCompressedFiles.html#compressed-content-cloudfront

For readers who come from google:
AWS S3 webp binary/octet-stream
RFC6838, RFC4855, RFC6838 - https://www.iana.org/assignments/media-types/media-types.xhtml

**Comment by michaelperret** [comment]
Apart from updating the underlying `mime.types` file is there any solution for this past specifying the `--content-type` flag? The guessed mime-type process is fantastic except for the unique case of woff and woff2 behind Cloudfront.

Are there any plans to address this @tim-finnigan?

**Comment by tim-finnigan** [comment]
@michaelperret I don't have any updates as far as plans at this time. There are workarounds as mentioned in this issue. The team is still tracking this feature request in their backlog and will prioritize addressing it accordingly.

**Comment by chasepo** [comment]
To resolve this, could we not just add a new argument to expose the existing functionality that leverages [mimetypes](https://docs.python.org/3/library/mimetypes.html) to look for and map filetypes? The `mimetypes` library already supports specifying ones own list of mimetype file paths, the CLI would then simply pass along the user provided path(s) into this when called internally. It would also provide an opportunity to document how this *magic* happens in the first place, rather than the same version of the cli behaving seemingly erratically depending on the host OS.

**Comment by timbru31** [comment]
Just an update for folks in 2025:
cpython recently had some [PRs](https://github.com/python/cpython/pull/129969) [merged](https://github.com/python/cpython/pull/20199) so hopefully with Python 3.14 this will auto solve some issues 🙏
